site/content/blog/2021-04-16-migrating-operator-custom-resource.md

```diff
@@ -0,0 +1,148 @@
+
+Unfortunately, conversion webhooks did not apply to our case as in addition to the change of version, we also changed both resource Group and Kind, effectively creating a whole new CR - `Mattermost`.
+
+
+#### Migrating to the new resource
+
+When the `ClusterInstallation` CR is created, the Mattermost Operator spins up a new instance of the application. This includes Kubernetes Deployments which are running the actual Mattermost application. When the CR is deleted all the resources are deleted together with it thanks to [Owner References](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#owners-and-dependents).
+
+This prevents us from simply running the script that would delete `ClusterInstallation` and create `Mattermost` in its place as it would cause the downtime of the Mattermost application (unless we chose to orphan all resources).
+
+To "exchange" the existing resources between different CRs as well as support both resources for some time, we decided to run two [Controllers](https://kubernetes.io/docs/concepts/architecture/controller/) as part of Mattermost Operator and make them perform the desired migration. As we wanted to retain some control over Mattermost instances that are being migrated we decided to introduce a new field to the `ClusterInstallation` Spec that would be a signal for the Operator to start the migration.
+
+```yaml
+apiVersion: mattermost.com/v1alpha1
+kind: ClusterInstallation
+metadata:
+  name: my-mattermost
+spec:
+  ...
+  migrate: "true"  # New field added to ClusterInstallation. Setting it to 'true' instructs the controller to start the migration.
+```
+
+This way we can perform the migration in the following way:
+- When the controller for `ClusterInstallation` sees that the `spec.migrate` is set to `true`, it stops regular reconciliation of the CR and starts the migration by converting the old resource to the new one. 
+- Immediately after the `Mattermost` resource is created, the controller for `Mattermost` sees it and starts to adjust existing resources like Services and Deployments to the new CR, making small changes and overriding owner references.
+- When we are sure that the controller for `Mattermost` successfully finished its work and Mattermost Pods are ready to serve traffic, the old `ClusterInstallation` is deleted, and voila!
+- As in the second phase the controller is just checking new Pods health, it can just do it once and requeue the resource for latter reconciliation while starting the migration for the next one, which gives us ability to migrate multiple resources at a time.
+
+The migration process has another advantage, that is if anything goes wrong it can be easily reverted just by setting `spec.migrate` back to `false` and removing the newly created `Mattermost` CR. The controller for `ClusterInstallation` will then claim the resources back and continue to monitor them.
```

Comment:
```suggestion
The migration process has another advantage: If anything goes wrong it can be easily reverted just by setting `spec.migrate` back to `false` and removing the newly created `Mattermost` CR. The controller for `ClusterInstallation` will then claim the resources back and continue to monitor them.
```